tests/test_binding.py

```diff
@@ -47,6 +49,63 @@ def __init__(self, v: int = 1) -> None:
         assert res.v == 1
 
 
+T_co = TypeVar("T_co", covariant=True)
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 10),
+    reason="TypeError: Parameters to generic types must be types.",
+)
+def test_bind_generic():
+    container = Container()
+    executor = SyncExecutor()
+    expected = 100
+
+    class GetterInterface(Protocol[T_co]):
+        @abstractmethod
+        def get(self) -> T_co:
+            ...
+
+    class GetterIntImpl(GetterInterface[int]):
+        def __init__(self, v: int) -> None:
+            self.v = v
+
+        def get(self) -> int:
+            return self.v
+
+    def factory() -> GetterIntImpl:
+        return GetterIntImpl(expected)
+
+    hook = bind_by_type(
+        Dependent(factory),
+        GetterInterface[int],
+    )
+    container.bind(hook)
+
+    # ===========================================
+    # clean `_tp_cache`
+    lru_cache_size = 128
+    _ = [GetterInterface[i] for i in range(lru_cache_size)]  # type: ignore[valid-type]
```

Comment:
You could just generate 128+ types by parametrizing, something like:

```python
unique_types = {
    Tuple[*perm] for n in range(1, 8)
    for comb in itertools.combinations([str, int, float, bool, None, object, Any], r=n)
    for perm in itertools.permutations(comb)
}
```

Reply:
> Python version 3.10 does not support starred expressions in subscriptions
Tuple[*perm]

Reply:
The last commit uses [private typing var](https://github.com/python/cpython/blob/edf701c4d83f399bff8612c34fef95da7663945e/Lib/typing.py#L245-L262).
What do you about it?
Looks like it is ok to use it here.

Reply:
I think it's fine